src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager.cs

```diff
@@ -131,57 +110,79 @@ public ConfigurationManager(string metadataAddress, IConfigurationRetriever<T> c
+                        CurrentConfiguration = await _configRetriever.GetConfigurationAsync(_metadataAddress, _docRetriever, CancellationToken.None).ConfigureAwait(false);
+                        Contract.Assert(CurrentConfiguration != null);
+                        _lastRefresh = now;
+                        _syncAfter = DateTimeUtil.Add(now.UtcDateTime, _automaticRefreshInterval);
+                    }
+                    catch (Exception ex)
+                    {
+                        _syncAfter = DateTimeUtil.Add(now.UtcDateTime, _automaticRefreshInterval < _refreshInterval ? _automaticRefreshInterval : _refreshInterval);
+                        if (CurrentConfiguration == null) // Throw an exception if there's no configuration to return.
+                            throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX20803, (_metadataAddress ?? "null")), ex));
+                        else
+                            LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX20806, (_metadataAddress ?? "null")), ex));
+                    }
+                }
+
+                // Stale metadata is better than no metadata
+                if (CurrentConfiguration != null)
+                    return CurrentConfiguration as T;
+                else
+                {
+                    throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX20803, (_metadataAddress ?? "null"))));
+                }
+            }
+            finally
+            {
+                _refreshLock.Release();
+            }
         }
 
         /// <summary>
         /// Obtains an updated version of Configuration.
         /// </summary>
         /// <param name="cancel">CancellationToken</param>
-        /// <returns>Configuration of type T.</returns>
-        /// <remarks>If the time since the last call is less than <see cref="AutomaticRefreshInterval"/> then <see cref="IConfigurationRetriever{T}.GetConfigurationAsync"/> is not called and the current Configuration is returned.</remarks>
-        public async Task<T> GetConfigurationAsync(CancellationToken cancel)
+        /// <returns>Configuration of type StandardConfiguration.</returns>
+        /// <remarks>If the time since the last call is less than <see cref="StandardConfigurationManager.AutomaticRefreshInterval"/> then <see cref="IConfigurationRetriever{T}.GetConfigurationAsync"/> is not called and the current Configuration is returned.</remarks>
+        public override async Task<StandardConfiguration> GetGeneralConfigurationAsync(CancellationToken cancel)
         {
             DateTimeOffset now = DateTimeOffset.UtcNow;
-            if (_currentConfiguration != null && _syncAfter > now)
+            if (UseCurrentConfiguration && CurrentConfiguration != null && _syncAfter > now)
```

Comment:
We can check for whether or not the CurrentConfiguration is valid inside of this method, but we cannot set this value until a successful validation has occurred (JsonWebTokenHandler will be responsible for this).